src/main/java/cd/go/contrib/elasticagents/dockerswarm/elasticagent/DockerService.java

```diff
@@ -69,55 +70,102 @@ public void terminate(DockerClient docker) throws DockerException, InterruptedEx
         try {
             LOG.debug("Terminating service " + this.name());
             docker.removeService(name);
+            deleteSecrets(docker, name);
         } catch (ServiceNotFoundException ignore) {
             LOG.warn("Cannot terminate a service that does not exist " + name);
         }
     }
 
+    private static void deleteSecrets(DockerClient docker, String serviceName) throws DockerException, InterruptedException {
+        docker.listSecrets().forEach(secret -> {
+            if (serviceName.equals(secret.secretSpec().labels().get(SWARM_SERVICE_NAME))) {
+                try {
+                    LOG.debug("Deleting secret " + secret.secretSpec().name());
+                    docker.deleteSecret(secret.id());
```

Comment:
Why are we deleting secrets? Isn't the point to just run `docker service create --secret src=SECRET_NAME,target=/path/to/file/in/container` and use an existing secret?

Won't this cause issues the next time some service wants to use the same secret?

Reply:
What if there is a change in secret, or deleted from profile ?

Reply:
The intention is that secrets will be provided outside of the plugin. So admins will be expected to upload the secrets out-of-band from the CLI or something similar. I'm not sure how you could possibly provide a plethora of secrets which can be several kilobytes in a profile in the cruise-config.xml. Some secrets will likely be shared between profiles so will cause duplication errors.

Reply:
Remember that binary blobs encoded as base64 will add about 33% to the size of the binary blob and add to the size of the config file.